T12548 make installer work from live systems

ramcq opened the pull request:
Assuming that the eoslive partition can be mounted by udisks, this allows the installer to detect if it is running in live mode and to install from an uncompressed image stored in endless/endless.img. The endless/eos-eosXX-etc-etc.img.asc signature is used for parsing the image version, personality, etc.

For https://phabricator.endlessm.com/T12548

ramcq committed:
check for endless.live_boot kernel parameter

The installer needs to behave differently (look for different paths and
partition names) in the case that the installer is running from a live system,
this is keyed off the kernel command line parameter endless.live_boot.

ramcq committed:
parse names from .img and .img.asc files as well as .gz and .xz

(Also fix a leak of 'product'.)

ramcq committed:
add support for installing uncompressed .img files

ramcq committed:
allow the signature file to be set independently from the image file

ramcq committed:
use "eoslive" partition when in live mode

This look for images on the eoslive partition in "endless" subdirectory when we
are running in live mode. Regular mode looks at the root folder of "eosimages"
as before.

wjt committed:
Look for endless.img when running in live mode

Provided there are no other images in the endless/ folder, when running in live
mode the installer will also accept an uncompressed image named
endless.img (which it will read from a mapped block device if available), find the
correct signature by reading the 'live' flag file, and
use the .img.asc signature file with its regular eos-eosXXX-etc-etc name
structure to parse out the image version, personality, etc.

wjt committed:
diskimage: const-ify some string parameters

wjt commented:
I have updated this PR with a number of cool new features:
1. it applies to `master`
2. it compiles
3. it (probably) works

(In particular, it deals with `endless.img` being hidden from `readdir` and being unreadable even if you find it.)

wjt committed:
polkit: grant 'live' user access if endless.live_boot-ed

polkit.spawn() is defined to throw an exception if the command returns
non-zero.

Grabbing the flag up-front and caching the result may be premature
optimization but it avoids having to contort the logic of the condition
to defer calling 'grep' until the last minute.

I don't see any security implications of adding this rule
unconditionally in all images. Even if someone creates a user called
'live' in a normal installed system, the endless.live_boot flag should
not be present. If it *is* present, then the user must have manually
edited the kernel command line, so granting a local user access to
udisks and co doesn't seem problematic.

https://phabricator.endlessm.com/T12548

wjt committed:
Install systemd units to the correct directory

They should be in /lib/systemd, not /etc.

ramcq commented on eos-installer-data/20-eos-installer.rules:
```diff
@@ -1,13 +1,23 @@
 
+var isLiveSystem;
+try {
+    polkit.spawn(['/bin/grep', '--quiet', 'endless.live_boot', '/proc/cmdline']);
```
Possibly excessively picky but, is it worth doing grep -e \bendless.live_boot\b ?

ramcq replied:
(Also endless.live_boot in any case.)

ramcq commented on gnome-image-installer/pages/diskimage/gis-diskimage-page.c:
```diff
@@ -242,16 +267,95 @@ static void add_image(GtkListStore *store, gchar *image)
+  gboolean is_live = gis_store_is_live_install ();
+
+  if (is_live)
+    imgpath = g_build_path ("/", path, "endless", NULL);
```
I guess this is maybe a little academic, but perhaps to "keep the meat from the flies" when we are acting on separate .img.*z{,.asc} files with the installer running under a live system, rather than reformatting the host with the running live system, maybe we should still look in the root folder and not do this "endless" subdir check any more?

wjt replied:
so you would have the layout be:

```
eoslive/
  eos-pqr.img.xz
  eos-pqr.img.xz.asc
  endless/
    endless.img
    eos-xyz.img.asc
    live  (contains "eos-xyz.img")
```

and you'd be offered both the pqr and xyz images?

ramcq commented on gnome-image-installer/pages/diskimage/gis-diskimage-page.c:
```diff
@@ -273,19 +377,30 @@ gis_diskimage_page_populate_model(GisPage *page, gchar *path)
 
+  if (is_live &&
+      /* we only use endless.img if there are no other images in the folder */
+      !gtk_tree_model_get_iter_first (GTK_TREE_MODEL (store), &iter) &&
```
Not sure whether this check adds much at this point, given we look for the image and signature specifically by name and there is no possible .asc ambiguity.

wjt replied:
Good point, I'll drop it.

ramcq commented on gnome-image-installer/pages/install/gis-install-page.c:
```diff
@@ -116,8 +120,17 @@ gis_install_page_prepare_read (GisPage *page, GError **error)
+      priv->decompressed = g_object_ref (input);
+    }
+
   g_object_unref (input);
```
I know I wrote this, but - could possibly avoid the ref/unref oddity by just moving the unref up into the if (priv->decompressor) path.

wjt replied:
huh, I see, can do.

wjt committed:
diskimage: begone, spurious whitespace

wjt committed:
diskimage: check unattended image config against live image

wjt commented:
I have fixed your various comments and also a couple of other things (see the latest 3 commits). I need to make some corresponding packaging changes. I also am not sure that the `eosimages.mount` stuff is actually needed given that we just use udisks to mount it, do you know the background on this?

wjt committed:
.desktop: make name suitable for human eyes

Previously, this name would never be displayed, but now it will appear
on the desktop of transient live sessions.

There is not enough space on our grid layout for “Reformat with Endless
OS” so we have to be a bit more terse.

https://phabricator.endlessm.com/T12548

wjt committed:
.gitignore: ignore tarballs
